apps/application-system/api/src/app/modules/application/lifecycle/application-lifecycle.service.ts

```diff
@@ -0,0 +1,158 @@
+import { Inject, Injectable } from '@nestjs/common'
+import { APPLICATION_CONFIG } from '../application.configuration'
+import type { ApplicationConfig } from '../application.configuration'
+import { Application } from '../application.model'
+import { ApplicationService } from '../application.service'
+import { AwsService } from '../files'
+import AmazonS3URI from 'amazon-s3-uri'
+import { LOGGER_PROVIDER } from '@island.is/logging'
+import type { Logger } from '@island.is/logging'
+
+export interface ApplicationPruning {
+  pruned: boolean
+  application: Pick<
+    Application,
+    'id' | 'attachments' | 'answers' | 'externalData'
+  >
+  failedAttachments: object
+}
+
+@Injectable()
+export class ApplicationLifeCycleService {
+  private processingApplications: ApplicationPruning[] = []
+
+  constructor(
+    @Inject(LOGGER_PROVIDER)
+    private logger: Logger,
+    private applicationService: ApplicationService,
+    private awsService: AwsService,
+  ) {
+    this.logger = logger.child({ context: 'ApplicationLifeCycleService' })
+  }
+
+  public async run() {
+    this.logger.info(`Starting application pruning...`)
+    await this.fetchApplicationsToBePruned()
+    await this.pruneAttachments()
+    await this.pruneApplicationData()
+    this.reportResults()
+    this.logger.info(`Application pruning done.`)
+  }
+
+  public getProcessingApplications() {
+    return this.processingApplications
+  }
+
+  private async fetchApplicationsToBePruned() {
+    const applications = (await this.applicationService.findAllDueToBePruned()) as Pick<
+      Application,
+      'id' | 'attachments' | 'answers' | 'externalData'
+    >[]
+
+    this.logger.info(`Found ${applications.length} applications to be pruned.`)
+    this.processingApplications = applications.map((application) => {
+      return {
+        pruned: true,
+        application,
+        failedAttachments: {},
+      }
+    })
+  }
+
+  private async pruneAttachments() {
+    for (const prune of this.processingApplications) {
+      const applicationAttachments = prune.application.attachments as {
+        key: string
+        name: string
+      }
+
+      const attachments = this.attachmentsToKeyArray(applicationAttachments)
+
+      if (attachments) {
+        for (const attachment of attachments) {
+          const { key, s3key, bucket, value } = attachment
+          try {
+            await this.awsService.deleteObject(bucket, s3key)
+
+            const exists = await this.awsService.fileExists(bucket, value)
```

Comment:
Hm. This feels redundant and expensive. Would be interested in knowing more about any false positive deletes you have experienced.